main.tf

```
@@ -106,15 +103,62 @@ locals {
       }
     }
   ]
+
+  # Check if any custom rule has tagStatus = "untagged"
+  has_custom_untagged_rule = length([
+    for rule in var.custom_lifecycle_rules : rule
+    if try(rule.selection.tagStatus, "") == "untagged"
+  ]) > 0
+
+  # Only include the default untagged rule if no custom untagged rule exists
+  final_untagged_image_rule = local.has_custom_untagged_rule ? [] : local.untagged_image_rule
+
+  # Prepare all rules that will be included in the policy before assigning priorities
+  all_lifecycle_rules = concat(
+    local.protected_tag_rules,
+    local.final_untagged_image_rule,
+    local.remove_old_image_rule,
+    var.custom_lifecycle_rules
+  )
+  any_tag_status_rules = [
+    for rule in local.all_lifecycle_rules : rule
+    if try(rule.selection.tagStatus, "") == "any"
+  ]
+  other_tag_status_rules = [
+    for rule in local.all_lifecycle_rules : rule
+    if try(rule.selection.tagStatus, "") != "any"
+  ]
+  # when we prioritize rules, we want to ensure that any tag status rules come last (e.g. lower priority)
+  sorted_lifecycle_rules = concat(local.other_tag_status_rules, local.any_tag_status_rules)
+
+  normalized_rules = [
+    for i, rule in local.sorted_lifecycle_rules : merge(
+      rule,
+      {
+        rulePriority = i + 1
+        selection = merge(
+          { for k, v in rule.selection : k => v if !(k == "tagPrefixList" || k == "tagPatternList") },
+          length(coalesce(lookup(rule.selection, "tagPrefixList", null), [])) > 0
+          ? { tagPrefixList = coalesce(lookup(rule.selection, "tagPrefixList", null), []) }
+          : {},
+          length(coalesce(lookup(rule.selection, "tagPatternList", null), [])) > 0
+          ? { tagPatternList = coalesce(lookup(rule.selection, "tagPatternList", null), []) }
+          : {}
+        )
```

```suggestion
        selection = rule.selection
```

I think we can skip this complexity if we have a validation rule that requires TagPrefixList or TagPatternList when required

---

Or replace that with skipping `null values` like

```
selection =  { for k, v in rule.selection : k => v if v!= null }
```

---

This is not the same, the hard thing is AWS requires a "any" tag_status to be last. If it is a "any" type then it cannot have `tagPrefixList` nor `tagPatternList`.